src/libcore/cell.rs

```diff
@@ -422,6 +511,24 @@ impl<T> Cell<T> {
     pub fn into_inner(self) -> T {
         unsafe { self.value.into_inner() }
     }
+
+    /// Private function used for implementing other methods when `T` is not `Copy`.
+    ///
+    /// This is `unsafe` because `left` and `right` may point to the same object.
+    #[unstable(feature = "move_cell", issue = "39264")]
+    unsafe fn take_and_restore<Result, F>(left: &Self, right: &Self, f: F) -> Result
+        where F: FnOnce(&T, &T) -> Result
+    {
+        let temp_left = mem::uninitialized();
+        let temp_right = mem::uninitialized();
+        let lhs = left.replace(temp_left);
+        let rhs = right.replace(temp_right);
+        // If panic happens in `f`, just let it crash.
```

Comment:
For the record: same UB issue here as in the `Clone` impl, which affects most of the other impls added in this PR.